File: src/dev_health_ops/core/taxonomy.py
"""Investment taxonomy helpers — pure business logic, no I/O.

Centralises the formatting and parsing of theme/subcategory keys so that
api/, workers/, and metrics/ code can share the same rules without importing
from each other.
"""

from __future__ import annotations

from dev_health_ops.investment_taxonomy import THEMES

__all__ = [
    "THEME_LABELS",
    "THEME_KEYS_BY_LABEL",
    "format_theme_label",
    "format_subcategory_label",
    "normalize_theme_key",
    "split_category_filters",
]

# Canonical human-readable labels for each theme key.
THEME_LABELS: dict[str, str] = {
    "feature_delivery": "Feature Delivery",
    "operational": "Operational / Support",
    "maintenance": "Maintenance / Tech Debt",
    "quality": "Quality / Reliability",
    "risk": "Risk / Security",
}

# Reverse mapping: lowercase label -> theme key.
THEME_KEYS_BY_LABEL: dict[str, str] = {
    label.lower(): key for key, label in THEME_LABELS.items()
}


def _title_case(value: str) -> str:
    return value.replace("_", " ").replace("-", " ").strip().title()


def format_theme_label(theme_key: str) -> str:
    """Return the canonical human-readable label for a theme key.

    Falls back to title-casing the raw key if not in the registry.
    """
    key = str(theme_key or "").strip().lower()
    if key in THEME_LABELS:
        return THEME_LABELS[key]
    return _title_case(theme_key)


def format_subcategory_label(subcategory_key: str) -> str:
    """Return a human-readable label for a subcategory key.

    E.g. "feature_delivery.customer" -> "Feature Delivery · Customer".
    """
    if "." not in subcategory_key:
        return _title_case(subcategory_key)
    theme, sub = subcategory_key.split(".", 1)
    return f"{_title_case(theme)} · {_title_case(sub)}"


def normalize_theme_key(theme_key: str | None) -> str | None:
    """Normalise a theme key or label to a canonical key, or return None."""
    if theme_key is None:
        return None
    raw = str(theme_key).strip()
    if not raw:
        return None
    lowered = raw.lower()
    if lowered in THEMES:
        return lowered
    if lowered in THEME_KEYS_BY_LABEL:
        return THEME_KEYS_BY_LABEL[lowered]
    return None


def split_category_filters(
    work_category: list[str | None] | None,
) -> tuple[list[str], list[str]]:
    """Split work category filter values into (themes, subcategories).

    A value with a dot (e.g. "feature_delivery.customer") is treated as a
    subcategory; its theme prefix is also added to the themes list.
    A value without a dot is treated as a theme only.

    Args:
        work_category: List of raw category strings, possibly containing None.

    Returns:
        (themes, subcategories) — both deduplicated and ordered by first appearance.
    """
    themes: list[str] = []
    subcategories: list[str] = []
    for category in work_category or []:
        if not category:
            continue
        category_str = str(category).strip()
        if not category_str:
            continue
        if "." in category_str:
            subcategories.append(category_str)
            themes.append(category_str.split(".", 1)[0])
        else:
            themes.append(category_str)
    return list(dict.fromkeys(themes)), list(dict.fromkeys(subcategories))

File: cmd/query-api/internal/investmentflow/taxonomy.go
// Taxonomy label/key helpers -- ports dev_health_ops/core/taxonomy.py's
// _title_case, format_theme_label, format_subcategory_label,
// normalize_theme_key and split_category_filters, the pure-logic pieces
// api/services/investment_flow.py imports directly. THEMES membership
// reuses internal/jobs/workgraph/units' own already-transcribed taxonomy
// port (SortedThemes/IsTheme, dev_health_ops/investment_taxonomy.py)
// rather than a second copy of that registry; THEME_LABELS has no
// existing Go port anywhere in this repo (grepped, see this package's own
// PR notes) so it is transcribed here, the same "definition, not a
// derived table" reasoning units.taxonomy.go's own doc comment gives for
// its two sorted slices.
package investmentflow

import (
	"strings"

	"github.com/full-chaos/dev-health-ops/internal/jobs/workgraph/units"
)

// themeLabels ports THEME_LABELS (core/taxonomy.py:22-28) verbatim.
var themeLabels = map[string]string{
	"feature_delivery": "Feature Delivery",
	"operational":      "Operational / Support",
	"maintenance":      "Maintenance / Tech Debt",
	"quality":          "Quality / Reliability",
	"risk":             "Risk / Security",
}

// themeKeysByLabel ports THEME_KEYS_BY_LABEL (core/taxonomy.py:31-33):
// the reverse mapping, lowercase label -> theme key.
var themeKeysByLabel = buildThemeKeysByLabel()

func buildThemeKeysByLabel() map[string]string {
	out := make(map[string]string, len(themeLabels))
	for key, label := range themeLabels {
		out[strings.ToLower(label)] = key
	}
	return out
}

// titleCase ports _title_case (core/taxonomy.py:36-37): Python's
// str.title() upper-cases the first letter of every "word" (a maximal
// run of alphabetic characters) and lower-cases every other letter in
// that run -- NOT the same as Go's strings.Title (deprecated, splits on
// whitespace only) or a naive per-space Title. strings.ToTitle-per-rune
// with a word-boundary tracker reproduces str.title()'s exact rule for
// the ASCII underscore/hyphen/space-delimited keys this taxonomy ever
// feeds it (confirmed: every SortedThemes/SortedSubcategories leaf is
// ASCII lowercase with '_' separators only).
func titleCase(value string) string {
	replaced := strings.ReplaceAll(strings.ReplaceAll(value, "_", " "), "-", " ")
	trimmed := strings.TrimSpace(replaced)
	if trimmed == "" {
		return trimmed
	}
	var b strings.Builder
	b.Grow(len(trimmed))
	prevIsLetter := false
	for _, r := range trimmed {
		isLetter := (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z')
		switch {
		case isLetter && !prevIsLetter:
			b.WriteRune(toUpperASCII(r))
		case isLetter:
			b.WriteRune(toLowerASCII(r))
		default:
			b.WriteRune(r)
		}
		prevIsLetter = isLetter
	}
	return b.String()
}

func toUpperASCII(r rune) rune {
	if r >= 'a' && r <= 'z' {
		return r - ('a' - 'A')
	}
	return r
}

func toLowerASCII(r rune) rune {
	if r >= 'A' && r <= 'Z' {
		return r + ('a' - 'A')
	}
	return r
}

// formatThemeLabel ports format_theme_label (core/taxonomy.py:40-48).
func formatThemeLabel(themeKey string) string {
	key := strings.ToLower(strings.TrimSpace(themeKey))
	if label, ok := themeLabels[key]; ok {
		return label
	}
	return titleCase(themeKey)
}

// formatSubcategoryLabel ports format_subcategory_label (core/taxonomy.py:
// 51-59).
func formatSubcategoryLabel(subcategoryKey string) string {
	idx := strings.Index(subcategoryKey, ".")
	if idx < 0 {
		return titleCase(subcategoryKey)
	}
	theme, sub := subcategoryKey[:idx], subcategoryKey[idx+1:]
	return titleCase(theme) + " · " + titleCase(sub)
}

// normalizeThemeKey ports normalize_theme_key (core/taxonomy.py:62-74).
// present is false for Python's None return (both the nil-input and the
// no-match cases collapse to that single Go zero value, matching every
// call site's own "empty means absent" usage -- drill_category and theme
// are always treated as optional strings, never distinguished from "not
// found" downstream).
func normalizeThemeKey(themeKey *string) (key string, present bool) {
	if themeKey == nil {
		return "", false
	}
	raw := strings.TrimSpace(*themeKey)
	if raw == "" {
		return "", false
	}
	lowered := strings.ToLower(raw)
	if units.IsTheme(lowered) {
		return lowered, true
	}
	if mapped, ok := themeKeysByLabel[lowered]; ok {
		return mapped, true
	}
	return "", false
}

// splitCategoryFilters ports split_category_filters (core/taxonomy.py:
// 77-105): a value with a dot is a subcategory (its theme prefix is also
// added to themes); a value with none is a theme. Both returned slices
// are deduplicated, ordered by first appearance -- Go's map-seen +
// append-in-order idiom reproduces Python's dict.fromkeys ordering
// exactly.
func splitCategoryFilters(workCategory []string) (themes, subcategories []string) {
	seenThemes := map[string]bool{}
	seenSubcategories := map[string]bool{}
	for _, category := range workCategory {
		trimmed := strings.TrimSpace(category)
		if trimmed == "" {
			continue
		}
		if idx := strings.Index(trimmed, "."); idx >= 0 {
			if !seenSubcategories[trimmed] {
				seenSubcategories[trimmed] = true
				subcategories = append(subcategories, trimmed)
			}
			theme := trimmed[:idx]
			if !seenThemes[theme] {
				seenThemes[theme] = true
				themes = append(themes, theme)
			}
		} else {
			if !seenThemes[trimmed] {
				seenThemes[trimmed] = true
				themes = append(themes, trimmed)
			}
		}
	}
	return themes, subcategories
}
